Develop wiregrid_actuator, wiregrid_kikusui / modify synacc to record synaccess switch on/off

sadachi5 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Developed new agents for wire grid loader:
1. wiregrid_kikusui
Rotate the wiregrid via KIKUSUI power supply

2. wiregrid_actuator
Load(insert) or unload(eject) the wiregrid via linear actuator
 
And modified existing agent:

3. synacc
 To record synaccess on/off. 

## Description
<!--- Describe your changes in detail -->
1. New agent
2. New agent
3. Added start_status_acq() function and self.switching variable to able to work with other functions.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
For grid loader

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Tested in Tokyo lab with 2nd grid loader.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

BrianJKoopman commented on agents/wiregrid_kikusui/src/moxaSerial.py:
```diff
@@ -0,0 +1,189 @@
+# 4/2009 BAS
```
Please replace the tabs (which seem to be only used in this file) with spaces.

BrianJKoopman commented on agents/wiregrid_kikusui/src/moxaSerial.py:
```diff
@@ -0,0 +1,189 @@
+
+	timeout = property(gettimeout,settimeout)
+
+def test1():
+	x = Serial_TCPServer(('google.com',80),timeout=1.15)
+	x.write('GET /\n')
+	print(x.readexactly(1000))
+
+def test2():
+	x = Serial_TCPServer(('google.com',80),timeout=0.15)
+	x.write('GET /\n')
+	print(x.read(10000))
```
Are these tests still needed? They look exploratory for behavior of `readexactly` and `read`. (And also they don't work if I run them...)

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+# Built-in python modules
+import time as tm
+import serial as sr
+import sys as sy
```
I'd probably drop the as xx here. These are all standard imports, and they're each only called once.

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+import sys as sy
+import os
+
+# Communication modules
+this_dir = os.path.dirname(__file__)
+sy.path.append(os.path.join(
+    this_dir, "..", "..","..", "MOXA"))
+import moxaSerial as mx  # noqa: E402
```
I'm not sure this path is valid. It would probably be better to move code that looks like it's going to be shared between this Agent and the hwp rotation agent (#157) to the core socs library (i.e. under `socs/socs/`), rather than manipulating the path manually here.

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+    Args:
+    rtu_port (str): Serial RTU port
+    tcp_ip (str): TCP IP address
+    tcp_port (int): TCP port
```
Missing the timeout argument in the docstring.

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+            else  :
+                self.clean_serial()
+                self.wait()
+                self.ser.write(str.encode("OUTP?\n\r"))
```
`\r\n` is the usual order of terminating characters, which would mean end of line. Maybe the Moxa spec has this flipped order, but thought I'd point it out as it stood out to me.

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+            print(
+                "Disconnecting from TCP IP %s at port %d"
+                % (self._tcp_ip, self._tcp_port))
+            pass
```
Unnecessary `pass`. All other `pass` statements in this file can also be removed.

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+        self.wait()
+        val = self.ser.readline()
+        msg = "Voltage set = %.3f V" % (float(val))
+        if (silent != True):
```
`if (silent != True):` -> `if not silent:`

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+        self.wait()
+        val = self.ser.readline()
+        msg = "Current set = %.3f A\n" % (float(val))
+        if (silent != True):
```
`if (silent != True):` -> `if not silent:`

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+
+    def turn_on(self,notmakesure=False):
+        """ Turn the PMX on """
+        if notmakesure == True:
```
Can shorten to `if notmakesure:`

BrianJKoopman commented on agents/wiregrid_kikusui/src/pmx.py:
```diff
@@ -0,0 +1,341 @@
+
+    def turn_off(self,notmakesure=False):
+        """ Turn the PMX off """
+        if notmakesure == True:
```
Can shorten to `if notmakesure:`

BrianJKoopman commented:
@sadachi5, sorry, still working through this. It's a lot of code for one review. Would you mind splitting off the new wiregrid_encoder Agent to a separate PR? (In retrospect I probably should have suggested these be split up initially, just to be more manageable on my end.)

sadachi5 commented:
> @sadachi5, sorry, still working through this. It's a lot of code for one review. Would you mind splitting off the new wiregrid_encoder Agent to a separate PR? (In retrospect I probably should have suggested these be split up initially, just to be more manageable on my end.)

It makes sense. I will remove the wiregrid_encoder agent in develop_wiregrid branch and
I will make a pull-request from the develop_wiregrid_encoder branch.

sadachi5 commented:
> > @sadachi5, sorry, still working through this. It's a lot of code for one review. Would you mind splitting off the new wiregrid_encoder Agent to a separate PR? (In retrospect I probably should have suggested these be split up initially, just to be more manageable on my end.)
> 
> It makes sense. I will remove the wiregrid_encoder agent in develop_wiregrid branch and I will make a pull-request from the develop_wiregrid_encoder branch.

I remove the wiregrid_encoder agent from this branch and created a new pull request for that agent in develop_wiregrid_encoder branch.

BrianJKoopman commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,603 @@
+                msg, v_val, c_val = self.cmd.user_input('VC?')
+                s_msg, s_val = self.cmd.user_input('O?')
+
+            self.log.info('Get voltage/current message: {}'.format(msg))
+            self.log.info('Get status message: {}'.format(s_msg))
+            return True,\
+                'Get Kikusui voltage / current: {} V / {} A [status={}]'\
+                .format(v_val, c_val, s_val)
```
This isn't the best place to pass that information. If you need the data available to an OCS Client, put it in `session.data`. If you need to record the values to disk for archiving (to .g3 file) publish them to a feed.

sadachi5 replied:
I removed the two self.log.info() printing lines. Is it what you want?
Here, I want just to show the volt, current, and status. (I don't need to record them.)
I think the return has a message about them. It is enough for me.

BrianJKoopman replied:
That's fine if they don't need to be recorded. I just wanted to make sure this didn't need to be acted on programmatically somewhere.

BrianJKoopman commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,603 @@
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import TimeoutLock
+
+# add PATH to ./src directory
+this_dir = os.path.dirname(__file__)
+sys.path.append(
+        os.path.join(this_dir, 'src'))
+
+# import classes
+import pmx
+import command
+from common import openlog, writelog
```
Resolved in #266.

BrianJKoopman commented on docs/agents/wiregrid_encoder.rst:
```diff
@@ -0,0 +1,101 @@
+.. highlight:: rst
```
Resolved in #266.

BrianJKoopman commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+     'arguments': [['--kikusui-ip', '10.10.10.71'],
+                   ['--kikusui-port', '29']]},
+
+ - kikusui-ip is an IP address of the serial-to-ethernet converter.
+ - kikusui-port is an asigned port for the KIKUSUI power supply.
+   (The converter has four D-sub ports to control 
+   multiple devices connected via serial communication. 
+   Communicating device is determined by the ethernet port number of the converter.)
```
Resolved in #266.

BrianJKoopman commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+---------
+
+.. autoclass:: agents.wiregrid_kikusui.kikusui_agent.KikusuiAgent
+    :members: set_on, set_off, set_v, set_c, get_vc,
+              calibration_wg, stepwise_rotation, IV_acq
```
Resolved in #266.

BrianJKoopman commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+ - stepwise_rotation(): 
+   Run step-wise rotation for wire-grid calibration.
+   In each step, seveal small-rotations are occurred to rotate 22.5-deg. 
+   - Parameters:
+       - feedback_steps: number of small rotations for each 22.5-deg step
+       - num_laps: number of laps (revolutions)
+       - stopped_time: stopped time [sec] for each 22.5-deg step
+       - feedback_time: calibration constants for the 22.5-deg rotation
```
Resolved in #266.

BrianJKoopman commented on agents/wiregrid_actuator/src/DigitalIO.py:
```diff
@@ -0,0 +1,196 @@
+    def set_onoff(self, onoff=0, io_name=None):
+        """
+        Set True/False (ON/OFF) for the digital IOs
+        If io_name=None, return a list of the ON/OFFs for all the IOs
+        If io_name is list, return a list of the ON/OFFs for asked IOs
+        If io_name is string (one IO), return one ON/OFF
```
Resolved in #266.

BrianJKoopman commented on docs/agents/wiregrid_actuator.rst:
```diff
@@ -0,0 +1,139 @@
+Agent API
+---------
+
+.. autoclass:: agents.wiregrid_actuator.wiregrid_actuator.WiregridActuatorAgent
+    :members: check_limitswitch, check_stopper, 
+              insert, eject, insert_homing, eject_homing, insert_test, eject_test, 
+              motor_on, motor_off, stop, release, reconnect, start_acq
```
Resolved in #266.

BrianJKoopman commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,649 @@
+import limitswitch_config
+import stopper_config
+
+class WiregridActuatorAgent:
```
Still needs a docstring.

sadachi5 replied:
I added it.

BrianJKoopman commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,649 @@
+        self.verbose       = verbose
+
+        agg_params = {'frame_length': 60}
+        self.agent.register_feed('WGActuator', record = True, agg_params = agg_params)
```
Still needs changing, but I made that comment in the new code too. Will mark this as resolved.

BrianJKoopman commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+
+from ocs import ocs_agent, site_config, client_t
+from ocs.ocs_twisted import TimeoutLock
+class KikusuiAgent:
```
Still needs a docstring.

sadachi5 replied:
I added it.

BrianJKoopman commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+        self.feedback_steps = params['feedback_steps']
+        self.num_laps = params['num_laps']
+        self.stopped_time = params['stopped_time']
+        self.feedback_time = params['feedback_time']
```
Highly encourage the use of param decorators here.

sadachi5 replied:
done

BrianJKoopman commented:
About the repeated code between this and Bryce's PR. moxaSerial is now used in yet another PR. I've gone ahead and moved the code from `moxaSerial.py` and `pmx.py` out of #157 and into #270.

I included instructions for updating code to use this shared version in that PR. Once everyone acknowledges the need for these changes are alright we should move forward with using this shared location for the code and remove any local instances in this PR.

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,649 @@
+
+
+    def stop(self, session, params=None):
+        self.log.warn('Try to stop and hold the actuator.')
```
This stop command is not used in a normal operation.
This command is assumed to be used when you want to stop the movement of the actuator in an emergency.
So, I assigned all the logs to warn or error.

BrianJKoopman replied:
Gotcha, I guess I'm thinking more in the programmatic sense that nothing here as actually gone wrong within the code. I don't feel strongly about needing it to be in `info` though, so that's fine.

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,649 @@
+                return False
+            # power off stopper
+            if not self.__stopper_off() : return False
+            return True
```
Is it solved already?

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+                writelog(logfile, 'ON', 0.70, self.get_position(self.position_path, self.open_trial, self.Deg))
+                self.rotate_alittle(ret, 0.70)
+                time.sleep(self.agent_interval+1.)
+                writelog(logfile, 'OFF', 0., self.get_position(self.position_path, self.open_trial, self.Deg))
```
I want an isolated log file for debugging.
I add a debug option to turn ON the log file. It is OFF nominally.
By the way, in calibrate_wg() function, the writelog()s always occur because the calibrate_wg() is a command just for commissioning of the rotation motor and the log file is necessary to adjust the rotation control parameters (self.feedback_time).

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+            for i in range(open_trial):
+                with open(position_path) as f:
+                    position_data = f.readlines()
+                    position = position_data[-1].split(' ')[1].replace('\n','')
```
This function needs to know the current position.
So it reads a position file which is updated by the wiregrid_encoder agent.
This is different file from the file used in writelog() function.

BrianJKoopman replied:
Marking this as resolved, continued relevant discussion in https://github.com/simonsobs/socs/pull/183#discussion_r831512850.

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+            self.log.warn('Failed to open ENCODER POSITION FILE')
+            pass
+
+        return int(position)*Deg
```
You're right. Thanks. I removed the Deg argument in _get_position() function

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+
+    args = parser.parse_args()
+
+    site_config.reparse_args(args, 'KikusuiAgent')
```
done

sadachi5 commented on agents/wiregrid_kikusui/src/command.py:
```diff
@@ -0,0 +1,179 @@
+                set_val = self._int(argv.pop(0))
+                if set_val is not None:
+                    del self._PMX
+                    self._PMX = px.PMX(set_val)
```
I confirmed that this bug was fixed.

sadachi5 commented on agents/wiregrid_kikusui/src/command.py:
```diff
@@ -0,0 +1,179 @@
+import sys as sy
+
+# KIKUSUI-power-suuply control modules
+import src.pmx as pmx
```
You mean that we don't need this import even if it hits the `pmx.PMX` line (line 144)?
If so, I can remote this import.

sadachi5 commented on agents/wiregrid_kikusui/src/command.py:
```diff
@@ -0,0 +1,179 @@
+# Built-in python modules
+import sys as sy
```
done

sadachi5 commented on agents/wiregrid_kikusui/src/command.py:
```diff
@@ -0,0 +1,179 @@
+            # Turn output state ON
+            elif cmd == self._cmds["set_on"]:
+                return self._PMX.turn_on(notmakesure=True)
+            # Turn output state OFF
```
fixed it

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,502 @@
+            # wait for getting control
+            ret = self.__wait_control()
+            if ret :
+                self.cmd.user_input('on')
```
fixed it

sadachi5 commented on agents/wiregrid_kikusui/src/command.py:
```diff
@@ -0,0 +1,179 @@
+            elif cmd == self._cmds["check_out"]:
+                return self._PMX.check_output()
+            # Turn output state ON
+            elif cmd == self._cmds["set_on"]:
```
I changed 'on' or 'off' in kikusui_agent.py to 'ON' or 'OFF'.

sadachi5 commented on agents/lakeshore240/Dockerfile:
```diff
@@ -2,7 +2,7 @@
 
 # Use socs base image
-FROM socs:latest
+FROM simonsobs/socs:latest
```
Sorry for that. I misunderstood it.
I should use 'socs' isn't it?

I checkout it to the previous (socs) commit. (git checkout d4ae11db89b068dd2abdab5ecccdfcd474df7624 Dockerfile)

sadachi5 commented on agents/synacc/Dockerfile:
```diff
@@ -1,7 +1,7 @@
 
 # Use socs base image
-FROM socs:latest
+FROM simonsobs/socs:latest
```
I also revert it to socs.

sadachi5 commented on agents/wiregrid_actuator/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM simonsobs/socs:latest
```
Done

sadachi5 commented on agents/wiregrid_actuator/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+RUN apt-get update \
+ && apt-get install -y apt-utils \
+ && apt-get install -y wget libavahi-common-dev libavahi-client-dev libavahi-core-dev libcap-dev libdaemon-dev avahi-daemon \
+ && wget http://www.galil.com/sw/pub/all/crypto/GALIL-GPG-KEY-E29D0E4B.asc \
+ && apt-key add GALIL-GPG-KEY-E29D0E4B.asc \
+ && wget http://www.galil.com/sw/pub/ubuntu/20.04/galil.list \
+ && cp galil.list /etc/apt/sources.list.d \
+ && apt-get update
```
I moved most of lines in wg-actuator-entrypoint.sh to Dockerfile.

sadachi5 commented on agents/wiregrid_actuator/src/Actuator.py:
```diff
@@ -0,0 +1,400 @@
+        self._cleanG()
+        print('Actuator:__del__(): '
+              'Successfully close the actuator controller.')
+        return True
```
removed the return

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+            # Wait for a second before moving
+            time.sleep(1)
+            # Moving commands
+            ret, msg = self._insert(920, 1.0)
+            if not ret:
+                msg = 'insert(): '\
+                      'ERROR!: Failed insert() in _insert(850,1.0) | {}'\
```
Fixed it

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+                }
+            }
+
+        Parameters:
+           Nothing
```
Added the description

sadachi5 commented on docker-compose.yml:
```diff
@@ -147,3 +147,24 @@ services:
+  # --------------------------------------------------------------------------
+  # Wiregrid KIKUSUI power supply agent
+  # --------------------------------------------------------------------------
+  ocs-wgkikusui-agent:
+    image: "ocs-wgkikusui-agent"
+    build: ./agents/wiregrid_kikusui/
+
+  # --------------------------------------------------------------------------
+  # Wiregrid encoder agent
+  # --------------------------------------------------------------------------
+  ocs-wgencoder-agent:
+    image: "ocs-wgencoder-agent"
+    build: ./agents/wiregrid_encoder/
+
+  # --------------------------------------------------------------------------
+  # Wiregrid actuator agent
+  # --------------------------------------------------------------------------
+  ocs-wgactuator-agent:
+    image: "ocs-wgactuator-agent"
+    build: ./agents/wiregrid_actuator/
```
Done

sadachi5 commented on agents/wiregrid_actuator/wg-actuator-entrypoint.sh:
```diff
@@ -0,0 +1,9 @@
+/etc/init.d/dbus start
+/etc/init.d/avahi-daemon start
+/usr/sbin/gcapsd &
+apt-get install gclib
```
I tried to do that as you said. However, I cannot find gclib_python.tar.gz after apt-get install gclib.

sadachi5 replied:
Thanks for your advice.
I can extract the necessary files from gclib package.
So I removed setup.py and gclib.py in wiregrid_actuator.

sadachi5 commented on agents/wiregrid_actuator/setup.py:
```diff
@@ -0,0 +1,11 @@
+# Setup script for gclib python wrapper
+# run 'python setup.py install' at console
+from distutils.core import setup
+setup(name='gclib',
+      version='1.0',
+      description='Python wrapper for Galil gclib',
+      author='Galil Motion Control',
+      author_email='softwaresupport@galil.com',
+      url='http://www.galil.com',
+      py_modules=['gclib'],
+      )
```
As I said above, I could not find the tar.gz file containing setup.py.

sadachi5 replied:
I removed it because I could extract it by using apt-get.

sadachi5 commented on agents/wiregrid_actuator/gclib.py:
```diff
@@ -0,0 +1,473 @@
+"""
+Python wrapper for Galil gclib.
+Contact softwaresupport@galil.com with questions, comments, and suggestions.
+"""
```
As I said above, I could not find the tar.gz file containing gclib.py.

sadachi5 replied:
Removed it.

sadachi5 commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+   :func: make_parser
+   :prog: python3 kikusui_agent.py
+
+Dependencies
+------------
+
+This agent depends on src/command.py, moxaSerial.py, and pmx.py.
+These scripts are also used in hwp_rotation agent.
+Therefore, they should be shared.
```
Done

sadachi5 commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+**Main Function (Stepwise Rotation)**
+ - stepwise_rotation(): 
+   Run step-wise rotation for wire-grid calibration.
+   In each step, seveal small-rotations are occurred to rotate 22.5-deg. 
```
Done

sadachi5 commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+
+Below is an example client to insert and eject the kikusui::
+
+    from ocs import matched_client
+    wgkikusui = matched_client.MatchedClient('wgkikusui', args=[])
```
Done

sadachi5 commented on docs/agents/wiregrid_kikusui.rst:
```diff
@@ -0,0 +1,161 @@
+it is controlled via serial-to-ethernet converter.
+The converter is linked to the KIKUSUI 
+via RS-232 (D-sub 9pin cable).
+The agent communicates with the converter via eternet.
```
Done

sadachi5 commented on docs/agents/wiregrid_actuator.rst:
```diff
@@ -0,0 +1,139 @@
+        image: simonsobs/ocs-wgactuator-agent:latest
+        hostname: ocs-docker
+        network_mode: "host"
+        depends_on:
+          - "crossbar"
```
Done

sadachi5 commented on docs/agents/wiregrid_actuator.rst:
```diff
@@ -0,0 +1,139 @@
+
+Below is an example client to insert and eject the actuator::
+
+    from ocs import matched_client
+    wgactuator = matched_client.MatchedClient('wgactuator', args=[])
```
Done

sadachi5 commented on docs/agents/wiregrid_actuator.rst:
```diff
@@ -0,0 +1,139 @@
+    from ocs import matched_client
+    wgactuator = matched_client.MatchedClient('wgactuator', args=[])
+
+    # Insert the actuator
+    wgactuator.insert.start()
+    wgactuator.insert.wait()
+
+    # Eject the actuator
+    wgactuator.eject.start()
+    wgactuator.eject.wait()
```
Done

sadachi5 commented on agents/wiregrid_actuator/src/DigitalIO.py:
```diff
@@ -0,0 +1,196 @@
+    """
+
+    def __init__(self, name, io_list, g,
+                 get_onoff_reverse=False, set_onoff_reverse=False, verbose=0):
```
done

sadachi5 commented on agents/wiregrid_actuator/src/DigitalIO.py:
```diff
@@ -0,0 +1,196 @@
+                      .format(self.io_names)\
+                    + 'DigitalIO[{}]:get_onoff():       '\
+                      'Asked IO names = {}'.format(self.name, io_name)
+                print(msg)
```
Removed all print statement before raising an error in DigitalIO.py, Actuator.py

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+                onoff_dict_ls = {}
+                onoff_dict_st = {}
+                # Get onoff
+                onoff_mt = self.actuator.get_motor_onoff()
```
I added try & exception statements. If it catches an error, it will return False and an error message.

sadachi5 commented on agents/wiregrid_actuator/src/DigitalIO.py:
```diff
@@ -0,0 +1,196 @@
+              '{}'.format(self.name, 'ON' if onoff else 'OFF', set_io_names))
+        for name in set_io_names:
+            self._set_onoff(onoff, name)
+        return True
```
Remove all `return True`

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+                  '--> Stop moving! | Exception = "{}"'.format(flabel, e)
+            self.log.error(msg)
+            return False, msg
+        # 2nd trial (double check)
+        try:
+            self.actuator.st.set_allon()
+        except Exception as e:
+            msg = '_insert_eject()[{}]: '\
+                  'ERROR!: Failed to run the stopper set_allon() '\
+                  '--> Stop moving! | Exception = "{}"'.format(flabel, e)
+            self.log.error(msg)
+            return False, msg
```
Added `get_onoff()` check

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+        if params is None:
+            params = {}
+        distance = params.get('distance', 10)
+        speedrate = params.get('speedrate', 10)
```
I modified the default value and docstring.

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+        if params is None:
+            params = {}
+        distance = params.get('distance', 10)
+        speedrate = params.get('speedrate', 10)
```
Done

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,603 @@
+        """
+        Set voltage [V]
+
+        Parameters:
+            volt: set voltage [V] (should be ONLY 12)
```
I modified the allowed value. (12V --> 0--12 V)

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,603 @@
+            logfile.close()
+            return True, 'Set Kikusui off'
+
+    def set_c(self, session, params=None):
```
I did

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,603 @@
+        self.kikusui_ip = kikusui_ip
+        self.kikusui_port = int(kikusui_port)
+
+        self.position_path = '/data/wg-data/position.log'
+        self.action_path = '/data/wg-data/action/'
```
I assume to mount a local directory on /data/wg-data in wiregrid_encoder and wiregrid_kikusui containers.
The position.log should be shared between the two agents because the kikusui agent needs to check the current position recorded by the encoder agent.
Maybe, there is a way to retrieve the current position recorded by the encoder agent. If we can do that, please let me know how to do that.

BrianJKoopman replied:
Yes, the way to do this is to use the [session.data](https://ocs.readthedocs.io/en/develop/developer/session-data.html). So the kikusui agent will need to create a client for the encoder agent, check the status of the `acq()` Process, and inspect the `session.data` object for the current position. (The encoder will of course have to put the position in `session.data`.)

The kikusui Agent will need to be told what the encoder agent's instance-id is, and you'll likely want to protect against the encoder agent not being online when the kikusui agent starts the client.

BrianJKoopman replied:
I believe this is resolved in the changes [here](https://github.com/simonsobs/socs/pull/183/files/a28bab332b63cdd72ebe6889edc02ea94f163403..2e4b112ceee85c5d6e72d459a9a4a8af138ba1ae).

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+
+        agg_params = {'frame_length': 60}
+        self.agent.register_feed(
+                'WGActuator', record=True, agg_params=agg_params)
```
Renamed it to wg_actuator

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+                'WGActuator', record=True, agg_params=agg_params)
+
+        try:
+            self.actuator = Actuator(
+                self.ip_address, sleep=self.sleep,
```
The controller is totally changed to a more sophisticated one. It can communicate via Ethernet!

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,845 @@
+
+    parser = make_parser()
+    args = site_config.parse_args(
+        agent_class='WGActuatorAgent', parser=parser)
```
done

sadachi5 replied:
I also changed the class name for wiregrid_kikusui:
class: KikusuiAgent --> WiregridKikusuiAgent
agent class : WGKikusuiAgent --> WiregridKikusuiAgent

sadachi5 commented on docs/agents/wiregrid_actuator.rst:
```diff
@@ -0,0 +1,139 @@
+These preparations are implemented in Dockerfile and wg-actuator-entrypoint.sh.
+Probably, the host machine requires the installation of gcapsd.
+Please install it via (in Ubuntu):
+    $ sudo apt install gcapsd
```
I added more descriptions of how to install them.
They follow the description in an official manual: 
[https://www.galil.com/sw/pub/all/doc/gclib/html/ubuntu.html](URL)

sadachi5 replied:
As you guess, `gclib` uses the `gcapsd` internally.
We start the `gcapsd` in L.5 `wiregrid_actuator/wg-actuator-entrypoint.sh`.
When we use this aganet, we mounted the /usr/sbin of the host PC, which has `gcapsd`, onto the /usr/sbin of the container in a ocs-site-config file.

This procedure works well. But, if you want, I'm OK to run this agent on the host PC directly.

sadachi5 replied:
I modified the Dockerfile:
(I comment-outed the old RUN lines and added new lines as you asked.)

````
FROM socs:latest

WORKDIR /app/socs/agents/wiregrid_actuator

#RUN apt-get update \ 
# && apt-get install -y apt-utils \
# && apt-get install -y wget libavahi-common-dev libavahi-client-dev libavahi-core-dev libcap-dev libdaemon-dev avahi-daemon \
# && wget http://www.galil.com/sw/pub/all/crypto/GALIL-GPG-KEY-E29D0E4B.asc \
# && apt-key add GALIL-GPG-KEY-E29D0E4B.asc \
# && wget http://www.galil.com/sw/pub/ubuntu/20.04/galil.list \
# && cp galil.list /etc/apt/sources.list.d \
# && apt-get update \
# && apt-get install -y --download-only gclib \
# && dpkg -x $(find / -name "gclib*.deb") / \
# && apt-get install -y gclib \
# && tar -xvf /usr/share/doc/gclib/src/gclib_python.tar.gz \
# && python3 setup.py install

RUN apt-get update \ 
 && apt-get install -y apt-utils \
 && apt-get install -y wget libavahi-common-dev libavahi-client-dev libavahi-core-dev libcap-dev libdaemon-dev avahi-daemon \
 && wget http://www.galil.com/sw/pub/all/crypto/GALIL-GPG-KEY-E29D0E4B.asc \
 && apt-key add GALIL-GPG-KEY-E29D0E4B.asc \
 && wget http://www.galil.com/sw/pub/ubuntu/20.04/galil.list \
 && cp galil.list /etc/apt/sources.list.d \
 && apt-get update \
 && apt-get install -y --download-only gclib \
 && dpkg -x $(find / -name "gclib*.deb") / \
 && apt-get install -y gclib \
 && apt-get install -y --download-only gcapsd \
 && dpkg -x $(find / -name "gcapsd*.deb") /tmp/gcapsd/ \
 && mv /tmp/gcapsd/usr/sbin/gcapsd /usr/sbin/ \
 && rm -r /tmp/gcapsd/ \
 && tar -xvf /usr/share/doc/gclib/src/gclib_python.tar.gz \
 && python3 setup.py install

COPY . .

ENTRYPOINT ["./wg-actuator-entrypoint.sh"]

CMD ["--site-hub=ws://crossbar:8001/ws", \
     "--site-http=http://crossbar:8001/call"]
````

After that, I stopped the ``gcapsd.service`` by ``sudo systemctl stop gcapsd.service``.
And I modified ocs-site-config:
(removed ``/usr/sbin`` mounting)

````
  ocs-wgactuator-agent:
    <<: *log-options
    image: "ocs-wgactuator-agent"
    build: ../../../socs_others/develop_wiregrid/agents/wiregrid_actuator/
    hostname: kyoto-docker
    network_mode: "host"
    depends_on:
      - "crossbar"
    volumes:
      - ${OCS_CONFIG_DIR}:/config:ro
      #- /usr/sbin:/usr/sbin
    command:
      - "--instance-id=wgactuator"
````

The build path contains the modified Dockerfile above.

I stopped the docker-agent by ``docker-compose down`` and removed the docker image clearly by ``docker image rm ocs-wgactuator-agent``.
After that, I reran the docker agents by ``docker-compose up -d``.
The wire-grid actuator agent successfully started up and connected to the actuator controller.

This process does not use the host PC's ``/usr/sbin`` and ``gcapsd`` service.
This is originally what we want to do.
Actually, this is super exciting for me!
Thanks a lot for your great advice!

I will modify the wiregrid_actuator.rst document as well.

BrianJKoopman replied:
That's great! Glad it worked!

BrianJKoopman reviewed (changes requested):
Thanks for the updates. There are a few things in the recent changes to address, besides the few lingering comments from past reviews (which I'll respond to individually.)

I merged the latest from the `develop` branch into here, so the shared copies of `pmx.py` and `moxaSerial.py` are available. See details in #270. Please replace your usage of those modules with the shared version and remove the versions you've committed here.

sadachi5 commented:
I modified `kikusui_agent.py` to use `socs/agents/pmx.py` instead of `agents/wiregrid_kikusui/src/pmx.py`.
In this modification, I also added several functions in `socs/agents/pmx.py` for wiregrid_kikusui.


> Thanks for the updates. There are a few things in the recent changes to address, besides the few lingering comments from past reviews (which I'll respond to individually.)
> 
> I merged the latest from the `develop` branch into here, so the shared copies of `pmx.py` and `moxaSerial.py` are available. See details in #270. Please replace your usage of those modules with the shared version and remove the versions you've committed here.

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,898 @@
+        # Get parameters
+        distance = params.get('distance')
+        speedrate = params.get('speedrate')
+        if speedrate < 0. or speedrate < 1.0:
```
Thanks. I removed the speedrate check in the function.

sadachi5 commented on agents/wiregrid_actuator/wiregrid_actuator.py:
```diff
@@ -0,0 +1,898 @@
+from ocs import ocs_agent
+from ocs import site_config
+from ocs.ocs_twisted import TimeoutLock
+from twisted.internet.defer import inlineCallbacks
```
Done

sadachi5 commented on agents/wiregrid_kikusui/kikusui_agent.py:
```diff
@@ -0,0 +1,616 @@
+
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import TimeoutLock
+from twisted.internet.defer import inlineCallbacks
```
Done

BrianJKoopman reviewed (changes requested):
Thanks for the updates. Just one last thing to fix with the new additions to `pmx.py`.

I'm also going to push a fix for the docs build failure.

sadachi5 commented on socs/agent/pmx.py:
```diff
@@ -38,24 +38,51 @@ def __del__(self):
-        val = float(self.ser.readline())
-        msg = "Measured voltage = %.3f V" % (val)
-        # print(msg)
+        try :
+            val = float(self.ser.readline())
+            msg = "Measured voltage = %.3f V" % (val)
+            # print(msg)
+        except  ValueError:
+            msg = 'WARNING! Could not get correct voltage value! | Response = "%s"' % (val)
+            val = -999
+            print(msg)
         return msg, val
```
I don't understand clearly.
There is an initialization of ``val`` in the except section.
Is it not enough?

By the way, I noticed that the ``val`` should be float type (-999.) instead of int() (-999).
So I changed several initial values to -999.

This exception can occur if the read string in ``self.ser.eadline()`` has a non-number character.
This can occur if the communication between the PC and KIKUSUI has some trouble. 
I hope this will not occur in real usage at the site. But, it may occur in the commissioning phase or lab. test.

BrianJKoopman replied:
> I don't understand clearly. There is an initialization of `val` in the except section. Is it not enough?

Just looking at this segment here, the initialization is on line 69, but `val` is used in the string formatting on line 68, which will fail with: `NameError: name 'val' is not defined.` Swapping these lines would work.

sadachi5 commented on socs/agent/pmx.py:
```diff
@@ -70,19 +97,75 @@ def check_voltage_current(self):
         # print(msg)
         return voltage, current
 
+    def check_voltagesetting(self):
```
Done

BrianJKoopman reviewed (approved):
Thanks for the all the updates! I think this is now ready! I handled a merge of the latest from the `develop` branch, and fixed some small things.
